BUG: Include python-including headers first

DWesl opened the pull request:
Ran into this trying to install scikit-image
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

DWesl committed:
BUG: Include python-including headers first

mattip commented:
Pleae add comments so we don’t undo this in a code refactor

charris commented:
Might just put `Python.h` up top, that way it will stay there after running `clang-format`.

DWesl committed:
STY: Add comments noting that Python.h should be included first.

DWesl committed:
BUG: Move Python.h include first and add comments explaining why it should be there.

DWesl committed:
STY: Include Python.h explicitly and add comments explaining why it's there.

DWesl commented:
I added an explicit Python.h include to those two files, with comments to explain why.

I then went a bit overboard and added those comments to every other Python.h include in the repository, moving it higher in the file if necessary.

ngoldbaum commented on numpy/_core/include/numpy/halffloat.h:
```diff
@@ -1,6 +1,9 @@
 #ifndef NUMPY_CORE_INCLUDE_NUMPY_HALFFLOAT_H_
 #define NUMPY_CORE_INCLUDE_NUMPY_HALFFLOAT_H_
 
+/* Any file that includes Python.h must include it before any other files */
+/* https://docs.python.org/3/extending/extending.html#a-simple-example */
+/* npy_common.h includes Python.h so it also counts in this list */
```
If we're going to touch every single header, I'd prefer it if this comment could be formatted like other comments in the numpy repo

```suggestion
/* 
   Any file that includes Python.h must include it before any other files
   https://docs.python.org/3/extending/extending.html#a-simple-example            
*/
```

I also think the addendum about `npy_common.h` is only relevant for files that include, and even then is maybe more confusing than helpful, since the most important thing is that `Python.h` comes first and the comment makes that clear.

DWesl replied:
> I also think the addendum about `npy_common.h` is only relevant for files that include it,

The intent was that `npy_common.h` is a file that includes Python.h, and so either it or Python.h needs to be the first include in any file that includes it, but that's not relevant for this batch of comments and I can see how that wasn't clear.

ngoldbaum commented:
Just a drive-by thought: maybe we could do this using a lint check instead of adding identical comments all over the codebase?

Unfortunately I let https://github.com/numpy/numpy/pull/28634 sit, otherwise we'd already have a C linter you could add to....

mattip commented:
>  Should I revert the last two comments, add that as a check, and adjust the files it flags?

Yes, I think that would be better. We have a linter check on [azure](https://github.com/numpy/numpy/blob/c96fcc8f7a5030098f8e0c304ce20864c3509021/azure-pipelines.yml#L55)

> should I move existing headers to the top of the files

I think that would be a less noisy change.

DWesl committed:
DEV: Add Python.h linter check to azure lint job

DWesl committed:
Revert "BUG: Move Python.h include first and add comments explaining why it should be there."

This reverts commit 83c46c9c187fc98c4bd057936390b97d6f7a77b2.

DWesl committed:
Revert "STY: Include Python.h explicitly and add comments explaining why it's there."

This reverts commit 79ab7673af83fc04cc9dfed74a10184661c159e1.

DWesl committed:
STY: Style fixes requested by linter

DWesl committed:
Revert "STY: Add comments noting that Python.h should be included first."

This reverts commit e7b47cf9762ca4b39d3179e2e900271c36cfa763.

mattip commented:
I am not sure this works. I reverted the change to `numpy/_core/src/common/blas_utils.h` and ran the script, it did not complain about the bad include order. Did I do something wrong?

DWesl committed:
BUG: Handle trailing comments on include lines in check_python_h_first.py

charris commented:
Note that if Python.h is present, clang-format will  put it first.

DWesl committed:
FIX: The leaf header would be the header just checked

Not whichever was last included

tylerjereddy commented on tools/get_submodule_paths.py:
```diff
@@ -0,0 +1,23 @@
+        submodule_paths = [os.path.join(root_directory, path) for path in
+                           submodule_paths]
+    # vendored with a script rather than via gitmodules
+    submodule_paths.append(os.path.join(root_directory, 'scipy/_lib/pyprima'))
```
Just a minor observation as I scan through my GitHub notifications--I recall suggesting/adding this line specifically for SciPy in some work with Lucas I think, so it can probably be removed here.

I'm a bit rusty on the NumPy vendoring situation--I suppose if NumPy vendors things without using git submodules there could be other relevant entries to add here?

DWesl committed:
DEV,ENH: Recognize vendored files from .gitattributes

DWesl committed:
DEV,ENH: Streamline the Python.h linter to reduce false positives.

Also allow passing a specific directory.

DWesl committed:
DEV,ENH: Get Python.h linter passing.

DWesl commented:
I wrote the initial version of `check_python_h_first.py`, but there's additions from other people in SciPy; the `get_submodule_paths.py` module is from someone else.  Should I add comments to that effect to both files, or revert to my initial version from there and add my changes from here?

seberg commented:
I dunno if it matters much, since I guess you mostly wrote it either way.  It seems to me we should just get it in, or does anyone of the original reviewers want to have another look?

ngoldbaum reviewed (commented):
Left a couple comments. The changes to the headers look much more minimal than the version of this PR I last looked at, so that's good.

I didn't try to understand the logic in the script. It seems pretty complicated? But if SciPy is using it, it's probably reasonably battle-tested, so I think I'm ok with including it.

DWesl commented:
> I didn't try to understand the logic in the script. It seems pretty complicated? 

A short version of the core function: for each file in the list, it checks each line in that file for `#include` and variants (space before `#`, space between `#` and `include`, double quotes or angle brackets).  If it finds `#include`:
1. If the include is of a file known not to include any other files, continue to the next line
2. If the include is not of `Python.h`, note it's including a non-Python header
3. If the include is of `Python.h`, or a file known to include `Python.h`, mark this file as one known to include `Python.h` and return the number of includes of non-python files it already found earlier in the file

If it reaches the end of the file without including any files, that file is added to the list of files known not to include any other file.

The rest of it is basically a wrapper to exclude files in submodules or vendored projects from the list of files, and to try to sort the list so headers appear before the files that include them, then to count the number of files that first include a non-`Python.h` file and exit with that status

> But if SciPy is using it, it's probably reasonably battle-tested, so I think I'm ok with including it.

I think I added it to SciPy a year or so back (after some searching, as scipy/scipy#20536), as a cheaper alternative to a Cygwin CI run to catch the most common failures: I think SciPy takes an hour to build, and likely longer to run the tests.

DWesl commented on azure-pipelines.yml:
```diff
@@ -56,6 +56,9 @@ stages:
       failOnStderr: true
+    - script: |
+        python tools/check_python_h_first.py
+      displayName: 'Check Python.h is first file included'
```
I did this up into a proper package with tests over at https://github.com/DWesl/check-python-h-first.  I did have to make changes to get all the tests to pass; the most relevant change was telling the script that a function starting with `Py` probably implies `Python.h` should have been included.

DWesl committed:
Merge branch 'main' of numpy/numpy into python-h-first

DWesl committed:
Merge branch 'main' of numpy/numpy into python-h-first
